gpumon/gpumon_interface.ml

```diff
@@ -12,60 +12,147 @@
+type nvidia_vgpu_metadata_list = nvidia_vgpu_metadata list
+[@@deriving rpcty]
+
+
+(** Error wrapper *)
+type gpu_errors =
+  | NvmlInterfaceNotAvailable
+  (** Exception raised when gpumon is unable to load the nvml nvidia library *)
+  | NvmlFailure of string
+  (** Exception raised by the c bindings to the nvml nvidia library*)
+  | Gpumon_failure
+  (** Default exception raised upon daemon failure *)
+[@@default Gpumon_failure]
+[@@deriving rpcty]
+
+exception Gpumon_error of gpu_errors
+
+(** Error handler *)
+module GpuErrors = Error.Make(struct
+    type t = gpu_errors
+    let t = gpu_errors
+  end)
+let gpu_err = GpuErrors.error
+
+(** Functor to autogenerate API calls *)
+module RPC_API(R : RPC) = struct
+  open R
+
+  let param = Param.mk
 
-(** Exception raised when gpumon is unable to load the nvml nvidia library *)
-exception NvmlInterfaceNotAvailable
-(** Exception raised by the c bindings to the nvml nvidia library*)
-exception NvmlFailure of string
+  let description =
+    Interface.{ name = "Gpumon"
+              ; namespace = None
+              ; description =
+                  [ "This interface is used by Xapi and Gpumon to monitor "
+                  ; "physical and virtual GPUs."]
+              ;
+                version=(1,0,0)
+              }
 
+  let implementation = implement description
+
+  (** common API call parameters *)
+
+  let debug_info_p = param ~description:
+      ["Uninterpreted string used for debugging."]
+      debug_info
+
+  let domid_p = param ~description:
+      ["Domain ID of the VM in which the vGPU(s) is running."]
+      domid
+
+  let pgpu_address_p = param ~description:
+      ["PCI bus ID of the pGPU in which the VM is currently running"
+      ;"in the form `domain:bus:device.function` PCI identifier."]
+      pgpu_address
+
+  let nvidia_pgpu_metadata_p = param ~description:
+      ["Metadata of Nvidia physical GPU."]
+      nvidia_pgpu_metadata
+
+  let nvidia_vgpu_metadata_p = param ~description:
+      ["Metadata of Nvidia virtual GPU."]
+      nvidia_vgpu_metadata
+
+  let nvidia_vgpu_metadata_list_p = param ~description:
+      ["Metadata list of Nvidia virtual GPU."]
+      nvidia_vgpu_metadata_list
+
+  let compatibility_p = param ~description:
+      ["Value indicating whether two or more GPUs are compatible with each other."]
+      compatibility
 
-module Nvidia = struct
   (** Compatibility checking interface for Nvidia vGPUs *)
+  module Nvidia = struct
+
+    let get_pgpu_metadata =
+      declare "get_pgpu_metadata"
+        ["Gets the metadata for a pGPU, given its address (PCI bus ID)."]
+        (debug_info_p @->  pgpu_address_p @-> returning nvidia_pgpu_metadata_p gpu_err )
+
+    let get_pgpu_vm_compatibility =
+      declare "get_pgpu_vm_compatibility"
+        ["Checks compatibility between a VM's vGPU(s) and another pGPU."]
+        (debug_info_p @->  pgpu_address_p @-> domid_p @-> nvidia_pgpu_metadata_p @-> returning compatibility_p gpu_err )
```

Comment:
Very long line; how about
```
( debug_info_p 
@->  pgpu_address_p 
@-> domid_p 
@-> nvidia_pgpu_metadata_p 
@-> returning compatibility_p gpu_err 
)
```

or similar? This also provides room for comments on parameters when needed.